sdks/python/apache_beam/runners/interactive/interactive_beam.py

```diff
@@ -0,0 +1,199 @@
+
+    When you re-run the pipeline from the line you just added, squares will
+    use the init_pcoll data cached so you can have an interactive experience.
+
+  Currently the implementation mainly watches for PCollection variables defined
+  in user code. A watchable can be a dictionary of variable metadata such as
+  locals(), a str name of a module, a module object or an instance of a class.
+  The variable can come from any scope even local variables in a method of a
+  class defined in a module.
+
+    Below are all valid:
+
+    watch(__main__)  # if import __main__ is already invoked
+    watch('__main__')  # does not require invoking import __main__ beforehand
+    watch(self)  # inside a class
+    watch(SomeInstance())  # an instance of a class
+    watch(locals())  # inside a function, watching local variables within
+  """
+  current_env().watch(watchable)
+
+
+def create_pipeline(runner=None, options=None, argv=None):
+  """Creates a pipeline with interactive runner by default.
+
+  You can use run_pipeline() provided within this module to execute the iBeam
+  pipeline with other runners.
+
+  Args:
+    runner (~apache_beam.runners.runner.PipelineRunner): An object of
+      type :class:`~apache_beam.runners.runner.PipelineRunner` that will be
+      used to execute the pipeline. For registered runners, the runner name
+      can be specified, otherwise a runner object must be supplied.
+    options (~apache_beam.options.pipeline_options.PipelineOptions):
+      A configured
+      :class:`~apache_beam.options.pipeline_options.PipelineOptions` object
+      containing arguments that should be used for running the Beam job.
+    argv (List[str]): a list of arguments (such as :data:`sys.argv`)
+      to be used for building a
+      :class:`~apache_beam.options.pipeline_options.PipelineOptions` object.
+      This will only be used if argument **options** is :data:`None`.
+
+  Raises:
+    ~exceptions.ValueError: if either the runner or options argument is not
+      of the expected type.
+  """
+  if not runner:
+    runner = interactive_runner.InteractiveRunner()
+  return beam.Pipeline(runner, options, argv)
+
+
+def visualize(pcoll):
+  """Visualizes a PCollection."""
+  # TODO(ningk)
+  pass
+
+
+def run_pipeline(pipeline, runner=None, options=None):
+  """Runs an iBeam pipeline with selected runner and options.
+
+  When you are done with prototyping in an interactive notebook env, use this
+  method to directly run the pipeline as a normal pipeline on your selected
+  runner.
+  """
+  # TODO(ningk)
+  pass
+
+
+def current_env(cache_manager=None):
+  """Gets current iBeam environment."""
+  global _ibeam_env
+  if not _ibeam_env:
+    _ibeam_env = InteractiveEnv(cache_manager)
+  return _ibeam_env
+
+
+def new_env(cache_manager=None):
+  """Creates a new iBeam environment to replace current one."""
+  global _ibeam_env
+  _ibeam_env = None
+  return current_env(cache_manager)
+
+
+class InteractiveEnv(object):
```

Comment:
Should we spell out InteractiveEnv to InteractiveEnvironment?

Reply:
Yes, we can. And I really want to hide this class from the interactive Beam (iBeam) user. I'll move everything about the environment to a different module.
This is supposed to be used by developers of iBeam internally.
iBeam users shouldn't use or care about what the implementation is.